helios-services/src/main/java/com/spotify/helios/master/ZooKeeperMasterModel.java

```diff
@@ -893,42 +940,176 @@ private RollingUpdateOp rollingUpdateDeploy(final ZooKeeperClient client,
+      if (isRolloutTimedOut(client, deploymentGroup)) {
+        return opFactory.error("timed out while retrieving job status", host,
+                               RollingUpdateError.TIMED_OUT_RETRIEVING_JOB_STATUS);
+      }
+
+      // We haven't detected any errors, so assume the agent will write the status soon.
+      return opFactory.yield();
+    }
+
+    if (!taskStatus.getState().equals(TaskStatus.State.STOPPED)) {
+      // job isn't stopped yet.
+
+      // We exceeded the configured deploy timeout, and this job is still not stopped
+      if (isRolloutTimedOut(client, deploymentGroup)) {
+        return rollingUpdateTimedoutError(
+            opFactory, host, jobId, taskStatus, TaskStatus.State.STOPPED);
+      }
+
+      return opFactory.yield();
+    }
+
+    return opFactory.nextTask();
+  }
+
+  private RollingUpdateOp rollingUpdateMarkUndeployed(final ZooKeeperClient client,
+                                                      final RollingUpdateOpFactory opFactory,
+                                                      final DeploymentGroup deploymentGroup,
+                                                      final String host) {
+    try {
+      final Node node = client.getNode(
+          Paths.statusDeploymentGroupRemovedHosts(deploymentGroup.getName()));
+      final int version = node.getStat().getVersion();
+      final List<String> hostsToUndeploy = Json.read(node.getBytes(), STRING_LIST_TYPE);
+
+      if (!hostsToUndeploy.removeAll(ImmutableList.of(host))) {
+        // Something already removed this host. Don't bother trying to update the removed hosts.
+        return opFactory.nextTask();
+      }
+      /*
+      The below check() avoids the following race:
+
+      1. Master A reads list of hosts X and Y
+      2. Master B reads list of hosts X and Y
+      3. Master A removes host X
+      4. Master B removes host Y
+      5. Master A commits list with Y and without X
+      6. Master B commits list with X and without Y
+
+      We would end up thinking we'd successfully removed both hosts, but in fact would have only
+      removed host Y.
+
+      The check() will cause this RollingUpdateOp's ZK transaction to fail without incrementing the
+      task index, so this RollingUpdateOp will be retried by the next master that gets to it.
+      */
+      return opFactory.nextTask(ImmutableList.of(
+          check(Paths.statusDeploymentGroupRemovedHosts(deploymentGroup.getName()), version),
+          set(Paths.statusDeploymentGroupRemovedHosts(deploymentGroup.getName()),
+              Json.asBytes(hostsToUndeploy))));
+    } catch (KeeperException | IOException e) {
+      // We can't read the list of hosts to undeploy. Move on.
+      log.error("unable to mark host undeployed after removal from deployment group: "
+                + "deployment-group={}, host={}", deploymentGroup.getName(), host, e);
+      return opFactory.nextTask();
```

Comment:
Do you expected this to happen? If not, maybe make it an `opFactory.error()` instead?

It's not obvious to me that carrying on like normal in case of errors results in correct behaviour.